res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3179 @@
+        }
+
+        const field = {
+            callback: callback,
+            byteOffset: byteOffset,
+            bitOffset: bitOffset,
+            bitLength: bitLength,
+            oldData: defaultOldData
+        };
+        this.fields.push(field);
+
+        return {
+            disconnect: () => {
+                this.fields = this.fields.filter((element) => {
+                    return element !== field;
+                });
+            }
+        };
+    }
+
+    handleInput(reportData) {
+        const view = new DataView(reportData);
+
+        for (const field of this.fields) {
+            const numBytes = Math.ceil(field.bitLength / 8);
+            let data;
+
+            // Little endianness is specified by the HID standard.
+            // The HID standard allows signed integers as well, but I am not aware
+            // of any HID DJ controllers which use signed integers.
+            if (numBytes === 1) {
+                data = view.getUint8(field.byteOffset);
+            } else if (numBytes === 2) {
+                data = view.getUint16(field.byteOffset, true);
+            } else if (numBytes === 3) {
+                data = view.getUint32(field.byteOffset, true) >>> 8;
+            } else if (numBytes === 4) {
+                data = view.getUint32(field.byteOffset, true);
+            } else {
+                throw Error("field bitLength must be between 1 and 32");
+            }
+
+            // The >>> 0 is required for 32 bit unsigned ints to not magically turn negative
+            // because all Numbers are really 32 bit signed floats. Because JavaScript.
+            data = ((data >> field.bitOffset) & (2 ** field.bitLength - 1)) >>> 0;
+
+            if (field.oldData !== data) {
+                field.callback(data);
+                field.oldData = data;
+            }
+        }
+    }
+}
+
+class HIDOutputReport {
+    constructor(reportId, length) {
+        this.reportId = reportId;
+        this.data = Array(length).fill(0);
+    }
+    send() {
+        controller.send(this.data, null, this.reportId);
+    }
+}
+
+/*
+ * Components library
+ */
+
+class Component {
+    constructor(options) {
+        if (options) {
+            Object.keys(options).forEach(function(key) {
+                if (options[key] === undefined) { delete options[key]; }
+            });
+            Object.assign(this, options);
+        }
+        this.outConnections = [];
+        if (typeof this.key === "string") {
+            this.inKey = this.key;
+            this.outKey = this.key;
+        }
+        if (this.unshift !== undefined && typeof this.unshift === "function") {
+            this.unshift();
+        }
+        this.shifted = false;
+        if (this.input !== undefined && typeof this.input === "function"
+            && this.inReport !== undefined && this.inReport instanceof HIDInputReport) {
+            this.inConnect();
+        }
+        this.outConnect();
```

Comment:
I'm not sure it would make sense to assert the arg list length in that context: if a component was to define extra args, out of the parent spec, it probable better we still call the function without argument, rather than not calling the `unshift` hook as it can lead to very strange behavior (suddenly, some part of the controller stay blocked in shifted behavior). Will remove redundant type checks though

Reply:
Well it depends on how rigid the API is. In case of the `input` handler for example, its better to write a function that takes the 5 args passed by the controller engine and ignore the trailing unimportant ones instead of writing a handler that just takes less arguments. In case of the `shift`/`unshift` functions, IMO a similar argument would make sense because there are invocations of that without arguments in componentsJS.
If you want to handle more arguments make them optional: `(function opt(a=undefined){...}).length` is still `0`.

Reply:
Just to clarify: So we agree that checking the length makes sense for ensuring passed functions work with the interface?

Reply:
If you think that makes more sense, then I'm happy to add it! I did add args list length asserting for all custom callbacks. Hopefully, that's now looking good :crossed_fingers: